README.md
# tbtail

`tbtail` is [Tinybird](https://tinybird.co)'s agent for ingesting log file data into Tinybird and making it available for exploration. Currently, it ingests all data into a string column of your data source. You'll have to extract each field, using JSONExtract function. You have detailed information about how you can do it in this [blog post](https://blog.tinybird.co/2021/01/28/nginx-log-analysis/).

`tbtail` is based on `clicktail` tool see [here](https://github.com/Altinity/clicktail).

## Supported Parsers

Unless our base tool `clicktail` supports several parsers, `tbtail` has only been tested with Nginx. Feel free to use other parsers if you feel confident enough.

The complete list of parsers can be found in the [`parsers/` directory](parsers/), but as of this writing there are:

- [ArangoDB](parsers/arangodb/)
- [MongoDB](parsers/mongodb/)
- [MySQL](parsers/mysql/)
- [PostgreSQL](parsers/postgresql/)
- [nginx](parsers/nginx/)
- [regex](parsers/regex/)
- [mysqlaudit](parsers/mysqlaudit/)

## Installation

There are three installation options for `tbtail`:
+ From [source code](#install-from-source-code)
+ From downloaded [`deb` package](#install-from-deb-package)
* Fron downloaded [`bin` package](#install-from-bin-package)

### Install from source code

1. Clone the [`tbtail`](https://github.com/tinybirdco/tbtail) and [`libtb-go`](https://github.com/tinybirdco/libtb-go) repo. 
2. Compile both using `go build`, `go install`, or create your makefile.

First of all, remember you should set your GOPATH var pointing to your go project folder.

Use go 1.14. It is not compatible with higher versions.

```
go get github.com/tinybirdco/tbtail
```

```
GOPATH=<go-project-folder> go build
GOPATH=<go-project-folder> go install
```

The binary will be installed into your bin folder `$GOPATH/bin/tbtail`

If you need to change anything highly probably you will need our `tblib-go` tool too.

```
go get github.com/tinybirdco/libtb-go
```

```
GOPATH=<go-project-folder> go build
GOPATH=<go-project-folder> go install
```

You can also use our [`Makefile`](./Makefile) to build, clean, install, and packaging the code.

- Use `make clean` to clean the project
- Use `make build` to build `tbtail`
- Use `make build-all` to build `tbtail` and `lib-tb`
- Use `make install` to install `tbtail`
- Use `make package` to build the .deb package

Remember that, in all cases, you need `GOPATH` to point to your root workspace.

### Releasing a new version

Everytime you create a new version remeber you need to bump the version in [version.txt](version.txt). It is important because the binary file, and the package file will contain the version you have generated.

Add all the changes to the [changelog](changelog.md) including the link to the binary file with the changes.

Do not forget to tag the code with the released version.

### Install from deb package

Download our latest package from [here](changelog.md).

Install the package as follows:

```sh
sudo dpkg -i tbtail_1.0.0_amd64.deb
```

After installing it you'll have:
- The binary file in `/usr/local/bin/tbtail`
- The configuration file in `/etc/tbtail/tbtail.conf`
- The service file in `/lib/systemd/system/tbtail.service`

You have detailed information about how to use the tool [here](#usage).

### Install from bin package

Download our latest package from [here](changelog.md).

You can place the binary file where you want and run it using the command line or a configuration file as explained in [Usage](#usage).

## Usage

Make sure you have created the data source in Tinybird that logs will be appended to. You can use the `nginx_raw.datasource` file: 

```sh
tb push datasource schema/tinybird/nginx_raw.datasource
tb push datasource schema/tinybird/nginx_log.datasource
tb push datasource schema/tinybird/nginx_mv.pipe
```

For a complete data project that will store the logs and materialize it with several granularity levels, you can check out [this project](https://github.com/tinybirdco/nginx-logs-analytics).

Once the Data Source is prepared to receive logs, you can run the binary as follows:

```sh
TB_WRITE_API_KEY=p.ey.. # token with either DATASOURCES:APPEND:nginx_raw, DATASOURCES:CREATE or ADMIN scope
tbtail --dataset nginx_raw --parser nginx --file /var/log/nginx/access.log --nginx.conf /etc/nginx/nginx.conf --nginx.format combined --writekey $TB_WRITE_API_KEY
```

You can also run the binary using a config file like [tbtail.conf](./tbtail.conf). This file contains a description for each available option.

If you have installed our package you can manage the service using `systemctl` command:  

```sh
sudo systemctl start tbtail
sudo systemctl stop tbtail
sudo systemctl restart tbtail
systemctl status tbtail
```

## Development

Make sure you have go 1.14. It is not compatible with higher versions.



changelog.md
# Changelog

Here you can find all the changes to this project. You can download either the `deb` package or our binary package for linux-64.

## [1.0.0] (2021-04-16)

This version has been tested with Nginx logs.

### Downloads

* [deb package](https://storage.googleapis.com/tbtail-releases/tbtail_1.0.0_amd64.deb)

* [binary file](https://storage.googleapis.com/tbtail-releases/tbtail_1.0.0_amd64)

### Fixes
- N/A


## [1.0.1] (2021-10-20)

This version is productive in all Tinybird servers with Nginx.

### Downloads

* [deb package](https://storage.googleapis.com/tbtail-releases/tbtail_1.0.0_amd64.deb)

* [binary file]()

### Fixes
- Changed append mode to use multipart instead of using the body of the request. This change has been made to avoid issues with the size of the request. Tinybird has the following [limits](https://docs.tinybird.co/api-reference/api-reference.html#limits-title)


## [1.1.0] (2022-02-11)

Added support for [HFI](https://docs.tinybird.co/api-reference/datasource-api.html#post-v0-events-title). Now you can decide to use HFI to send data or our regular API. If you want to use the HFI ingestion you need to run the tool using the `--hfi` flag. 

It is important to have in mind the HFI endpoint is intended to send events with a high rate at a small volume. Thus:
- Use low values in `send_frequency_ms`. Expected values will be around 500 - 1500 ms.
- Use low values in `send_bathc_size`. The HFI endopoint is not intended to be used with high data sizes so use a small number of events. In general, num_events * size_event < 

### Fixes
- Added support for [HFI](https://docs.tinybird.co/api-reference/datasource-api.html#post-v0-events-title).


## [1.1.0] (2022-02-17)

Added a new binary file for macOS (x86_64) and Windows (x86_64).

### Downloads

* [deb package](https://storage.googleapis.com/tbtail-releases/tbtail_1.1.0_amd64.deb)

* [binary file Linux](https://storage.googleapis.com/tbtail-releases/tbtail_1.1.0_linux_amd64)

* [binary file macOS](https://storage.googleapis.com/tbtail-releases/tbtail_1.1.0_macos_amd64)

* [binary file Windows](https://storage.googleapis.com/tbtail-releases/tbtail_1.1.0_windows_amd64.exe)

### Fixes
- N/A


pkg-test/test.sh
#!/bin/sh

# Smoke-test package installation by installing packages into a container. This
# assumes that packages exist in $GOPATH/bin

set -e

if [ "$#" -ne 1 ]; then echo "usage: test.sh <build id>"; exit 1; fi

BUILDID=$1
DEB=tbtail_${BUILDID}_amd64.deb
# RPM=tbtail-${BUILDID}-1.x86_64.rpm
DIR=`dirname $0`
echo docker build --build-arg package=$DEB -f Dockerfile.deb $DIR

cp $GOPATH/bin/$DEB $DIR
# cp $GOPATH/bin/$RPM $DIR
docker build --build-arg package=$DEB -f $DIR/Dockerfile.deb $DIR
# docker build --build-arg package=$RPM -f $DIR/Dockerfile.rpm $DIR


response_stats.go
package main

import (
	"strings"
	"sync"
	"time"

	"github.com/Sirupsen/logrus"
	"github.com/honeycombio/honeytail/event"
	"github.com/tinybirdco/libtb-go"
)

// responseStats is a container for collecting statistics about events sent
// via libtb. It counts interesting aspects of the events it gets and
// presents them for printing whenever it's called.
//
// the intent is to periodically print and flush the counters, eg once/minute

type responseStats struct {
	lock *sync.Mutex

	count       int
	statusCodes map[int]int
	bodies      map[string]int
	errors      map[string]int
	maxDuration time.Duration
	sumDuration time.Duration
	minDuration time.Duration
	event       *event.Event

	totalCount       int
	totalStatusCodes map[int]int
}

// newResponseStats initializes the struct's complex data types
func newResponseStats() *responseStats {
	r := &responseStats{}
	r.totalStatusCodes = make(map[int]int)
	r.lock = &sync.Mutex{}
	r.reset()
	return r
}

// update adds a response into the stats container
func (r *responseStats) update(rsp libtb.Response) {
	r.lock.Lock()
	defer r.lock.Unlock()
	r.count += 1
	r.statusCodes[rsp.StatusCode] += 1
	r.bodies[strings.TrimSpace(string(rsp.Body))] += 1
	if rsp.Err != nil {
		r.errors[rsp.Err.Error()] += 1
	}
	if r.minDuration == 0 {
		r.minDuration = rsp.Duration
	}
	if rsp.Duration < r.minDuration {
		r.minDuration = rsp.Duration
	} else if rsp.Duration > r.maxDuration {
		r.maxDuration = rsp.Duration
	}
	r.sumDuration += rsp.Duration
	ev := rsp.Metadata.(event.Event)
	r.event = &ev
}

// log the current stats and reset them all to zero.
// thread safe.
func (r *responseStats) logAndReset() {
	r.lock.Lock()
	defer r.lock.Unlock()
	r.log()
	r.reset()
}

// log the current statistics to logrus.
// NOT thread safe.
func (r *responseStats) log() {
	var avg time.Duration
	if r.count != 0 {
		avg = r.sumDuration / time.Duration(r.count)
	} else {
		avg = 0
	}
	logrus.WithFields(logrus.Fields{
		"count":            r.count,
		"lifetime_count":   r.totalCount + r.count,
		"slowest":          r.maxDuration,
		"fastest":          r.minDuration,
		"avg_duration":     avg,
		"count_per_status": r.statusCodes,
		"response_bodies":  r.bodies,
		"errors":           r.errors,
	}).Info("Summary of sent events")
	if r.event != nil {
		fields := make(map[string]interface{})
		fields["event"] = r.event.Data
		fields["event_timestamp"] = r.event.Timestamp
		logrus.WithFields(fields).Info("Last parsed event")
	}
}

// log the total count on its own
func (r *responseStats) logFinal() {
	r.lock.Lock()
	defer r.lock.Unlock()
	r.totalCount += r.count
	for code, count := range r.statusCodes {
		r.totalStatusCodes[code] += count
	}
	logrus.WithFields(logrus.Fields{
		"total attempted sends":               r.totalCount,
		"number sent by response status code": r.totalStatusCodes,
	}).Info("Total number of events sent")
}

// reset the counters to zero.
// NOT thread safe
func (r *responseStats) reset() {
	r.totalCount += r.count
	for code, count := range r.statusCodes {
		r.totalStatusCodes[code] += count
	}
	r.count = 0
	r.statusCodes = make(map[int]int)
	r.bodies = make(map[string]int)
	r.errors = make(map[string]int)
	r.maxDuration = 0
	r.sumDuration = 0
	r.minDuration = 0
}


leash.go
package main

import (
	"context"
	"crypto/sha256"
	"fmt"
	"math/rand"
	"os"
	"os/signal"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"sync"
	"syscall"
	"time"

	"github.com/Sirupsen/logrus"
	"github.com/honeycombio/dynsampler-go"
	"github.com/honeycombio/urlshaper"
	"github.com/tinybirdco/libtb-go"

	"github.com/Altinity/clicktail/parsers/mysql"
	"github.com/Altinity/clicktail/parsers/mysqlaudit"
	"github.com/honeycombio/honeytail/event"
	"github.com/honeycombio/honeytail/parsers"
	"github.com/honeycombio/honeytail/parsers/arangodb"
	"github.com/honeycombio/honeytail/parsers/htjson"
	"github.com/honeycombio/honeytail/parsers/keyval"
	"github.com/honeycombio/honeytail/parsers/mongodb"
	"github.com/honeycombio/honeytail/parsers/nginx"
	"github.com/honeycombio/honeytail/parsers/postgresql"
	"github.com/honeycombio/honeytail/parsers/regex"
	"github.com/honeycombio/honeytail/tail"
)

// actually go and be leashy
func run(options GlobalOptions) {
	logrus.Info("Starting tbtail")

	stats := newResponseStats()

	sigs := make(chan os.Signal, 1)
	ctx, cancel := context.WithCancel(context.Background())
	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)

	// spin up our transmission to send events to ClickHouse
	libhConfig := libtb.Config{
		WriteKey:             options.Reqs.WriteKey,
		Dataset:              options.Reqs.Dataset,
		APIHost:              options.APIHost,
		Hfi:                  options.Hfi,
		MaxConcurrentBatches: options.NumSenders,
		SendFrequency:        time.Duration(options.BatchFrequencyMs) * time.Millisecond,
		MaxBatchSize:         options.BatchSize,
		// block on send should be true so if we can't send fast enough, we slow
		// down reading the log rather than drop lines.
		BlockOnSend: true,
		// block on response is true so that if we hit rate limiting we make sure
		// to re-enqueue all dropped events
		BlockOnResponse: true,

		// limit pending work capacity so that we get backpressure from libtb
		// and block instead of sleeping inside sendTolibtb.
		PendingWorkCapacity: 20 * options.NumSenders,
	}
	if err := libtb.Init(libhConfig); err != nil {
		logrus.WithFields(logrus.Fields{"err": err}).Fatal(
			"Error occured while spinning up Transimission")
	}

	// compile the prefix regex once for use on all channels
	var prefixRegex *parsers.ExtRegexp
	if options.PrefixRegex == "" {
		prefixRegex = nil
	} else {
		prefixRegex = &parsers.ExtRegexp{regexp.MustCompile(options.PrefixRegex)}
	}

	// get our lines channel from which to read log lines
	var linesChans []chan string
	var err error
	tc := tail.Config{
		Paths:   options.Reqs.LogFiles,
		Type:    tail.RotateStyleSyslog,
		Options: options.Tail,
	}
	if options.TailSample {
		linesChans, err = tail.GetSampledEntries(ctx, tc, options.SampleRate)
	} else {
		linesChans, err = tail.GetEntries(ctx, tc)
	}
	if err != nil {
		logrus.WithFields(logrus.Fields{"err": err}).Fatal(
			"Error occurred while trying to tail logfile")
	}

	// set up our signal handler and support canceling
	go func() {
		sig := <-sigs
		fmt.Fprintf(os.Stderr, "Aborting! Caught signal \"%s\"\n", sig)
		fmt.Fprintf(os.Stderr, "Cleaning up...\n")
		cancel()
		// and if they insist, catch a second CTRL-C or timeout on 10sec
		select {
		case <-sigs:
			fmt.Fprintf(os.Stderr, "Caught second signal... Aborting.\n")
			os.Exit(1)
		case <-time.After(10 * time.Second):
			fmt.Fprintf(os.Stderr, "Taking too long... Aborting.\n")
			os.Exit(1)
		}
	}()

	// for each channel we got back from tail.GetEntries, spin up a parser.
	parsersWG := sync.WaitGroup{}
	responsesWG := sync.WaitGroup{}
	for _, lines := range linesChans {
		// get our parser
		parser, opts := getParserAndOptions(options)
		if parser == nil {
			logrus.WithFields(logrus.Fields{"parser": options.Reqs.ParserName}).Fatal(
				"Parser not found. Use --list to show valid parsers")
		}

		// and initialize it
		if err := parser.Init(opts); err != nil {
			logrus.Fatalf(
				"Error initializing %s parser module: %v", options.Reqs.ParserName, err)
		}

		// create a channel for sending events into libtb
		toBeSent := make(chan event.Event, options.NumSenders)
		doneSending := make(chan bool)

		// two channels to handle backing off when rate limited and resending failed
		// send attempts that are recoverable
		toBeResent := make(chan event.Event, 2*options.NumSenders)
		// time in milliseconds to delay the send
		delaySending := make(chan int, 2*options.NumSenders)

		// apply any filters to the events before they get sent
		modifiedToBeSent := modifyEventContents(toBeSent, options)

		realToBeSent := make(chan event.Event, 10*options.NumSenders)
		go func() {
			wg := sync.WaitGroup{}
			for i := uint(0); i < options.NumSenders; i++ {
				wg.Add(1)
				go func() {
					for ev := range modifiedToBeSent {
						realToBeSent <- ev
					}
					wg.Done()
				}()
			}
			wg.Wait()
			close(realToBeSent)
		}()

		// start up the sender. all sources are either sampled when tailing or in-
		// parser, so always tell libtb events are pre-sampled
		go sendToLibhoney(ctx, realToBeSent, toBeResent, delaySending, doneSending)

		// start a goroutine that reads from responses and logs.
		responses := libtb.Responses()
		responsesWG.Add(1)
		go func() {
			handleResponses(responses, stats, toBeResent, delaySending, options)
			responsesWG.Done()
		}()

		parsersWG.Add(1)
		go func(plines chan string) {
			// ProcessLines won't return until lines is closed
			parser.ProcessLines(plines, toBeSent, prefixRegex)
			// trigger the sending goroutine to finish up
			close(toBeSent)
			// wait for all the events in toBeSent to be handed to libtb
			<-doneSending
			parsersWG.Done()
		}(lines)
	}
	parsersWG.Wait()
	// tell libtb to finish up sending events
	libtb.Close()
	// print out what we've done one last time
	responsesWG.Wait()
	stats.log()
	stats.logFinal()

	// Nothing bad happened, yay
	logrus.Info("Tbtail is all done, goodbye!")
}

// getParserOptions takes a parser name and the global options struct
// it returns the options group for the specified parser
func getParserAndOptions(options GlobalOptions) (parsers.Parser, interface{}) {
	var parser parsers.Parser
	var opts interface{}
	switch options.Reqs.ParserName {
	case "regex":
		parser = &regex.Parser{}
		opts = &options.Regex
		opts.(*regex.Options).NumParsers = int(options.NumSenders)
	case "nginx":
		parser = &nginx.Parser{}
		opts = &options.Nginx
		opts.(*nginx.Options).NumParsers = int(options.NumSenders)
	case "json":
		parser = &htjson.Parser{}
		opts = &options.JSON
		opts.(*htjson.Options).NumParsers = int(options.NumSenders)
	case "keyval":
		parser = &keyval.Parser{}
		opts = &options.KeyVal
		opts.(*keyval.Options).NumParsers = int(options.NumSenders)
	case "mongo", "mongodb":
		parser = &mongodb.Parser{}
		opts = &options.Mongo
		opts.(*mongodb.Options).NumParsers = int(options.NumSenders)
	case "mysql":
		parser = &mysql.Parser{
			SampleRate: int(options.SampleRate),
		}
		opts = &options.MySQL
		opts.(*mysql.Options).NumParsers = int(options.NumSenders)
	case "mysqlaudit":
		parser = &mysqlaudit.Parser{}
		opts = &options.MySQLAudit
		opts.(*mysqlaudit.Options).NumParsers = int(options.NumSenders)
	case "postgresql":
		opts = &options.PostgreSQL
		parser = &postgresql.Parser{}
	case "arangodb":
		parser = &arangodb.Parser{}
		opts = &options.ArangoDB
	}
	parser, _ = parser.(parsers.Parser)
	return parser, opts
}

// modifyEventContents takes a channel from which it will read events. It
// returns a channel on which it will send the munged events. It is responsible
// for hashing or dropping or adding fields to the events and doing the dynamic
// sampling, if enabled
func modifyEventContents(toBeSent chan event.Event, options GlobalOptions) chan event.Event {
	// parse the addField bit once instead of for every event
	parsedAddFields := map[string]string{}
	for _, addField := range options.AddFields {
		splitField := strings.SplitN(addField, "=", 2)
		if len(splitField) != 2 {
			logrus.WithFields(logrus.Fields{
				"add_field": addField,
			}).Fatal("unable to separate provided field into a key=val pair")
		}
		parsedAddFields[splitField[0]] = splitField[1]
	}
	// do all the advance work for request shaping
	shaper := &requestShaper{}
	if len(options.RequestShape) != 0 {
		shaper.pr = &urlshaper.Parser{}
		if options.ShapePrefix != "" {
			shaper.prefix = options.ShapePrefix + "_"
		}
		for _, rpat := range options.RequestPattern {
			pat := urlshaper.Pattern{Pat: rpat}
			if err := pat.Compile(); err != nil {
				logrus.WithField("request_pattern", rpat).WithError(err).Fatal(
					"Failed to compile provided pattern.")
			}
			shaper.pr.Patterns = append(shaper.pr.Patterns, &pat)
		}
	}
	// initialize the dynamic sampler
	var sampler dynsampler.Sampler
	if len(options.DynSample) != 0 {
		sampler = &dynsampler.AvgSampleWithMin{
			GoalSampleRate:    options.GoalSampleRate,
			ClearFrequencySec: options.DynWindowSec,
			MinEventsPerSec:   options.MinSampleRate,
		}
		if err := sampler.Start(); err != nil {
			logrus.WithField("error", err).Fatal("dynsampler failed to start")
		}
	}
	// ok, we need to munge events. Sing up enough goroutines to handle this
	newSent := make(chan event.Event, options.NumSenders)
	go func() {
		wg := sync.WaitGroup{}
		for i := uint(0); i < options.NumSenders; i++ {
			wg.Add(1)
			go func() {
				for ev := range toBeSent {
					// do dropping
					for _, field := range options.DropFields {
						delete(ev.Data, field)
					}
					// do scrubbing
					for _, field := range options.ScrubFields {
						if val, ok := ev.Data[field]; ok {
							// generate a sha256 hash and use the base16 for the content
							newVal := sha256.Sum256([]byte(fmt.Sprintf("%v", val)))
							ev.Data[field] = fmt.Sprintf("%x", newVal)
						}
					}
					// do adding
					for k, v := range parsedAddFields {
						ev.Data[k] = v
					}
					// do request shaping
					for _, field := range options.RequestShape {
						shaper.requestShape(field, &ev, options)
					}
					// do dynsampling last so it can use request shaped fields
					if sampler == nil {
						ev.SampleRate = int(options.SampleRate)
					} else {
						key := makeDynsampleKey(&ev, options)
						sr := sampler.GetSampleRate(key)
						if rand.Intn(sr) != 0 {
							ev.SampleRate = -1
						} else {
							ev.SampleRate = sr
						}
					}
					newSent <- ev
				}
				wg.Done()
			}()
		}
		wg.Wait()
		close(newSent)
	}()
	return newSent
}

// makeDynsampleKey pulls in all the values necessary from the event to create a
// key for dynamic sampling
func makeDynsampleKey(ev *event.Event, options GlobalOptions) string {
	key := make([]string, len(options.DynSample))
	for i, field := range options.DynSample {
		if val, ok := ev.Data[field]; ok {
			switch val := val.(type) {
			case bool:
				key[i] = strconv.FormatBool(val)
			case int64:
				key[i] = strconv.FormatInt(val, 10)
			case float64:

				key[i] = strconv.FormatFloat(val, 'E', -1, 64)
			case string:
				key[i] = val
			default:
				key[i] = "" // skip it
			}
		}
	}
	return strings.Join(key, "_")
}

// requestShaper holds the bits about request shaping that want to be
// precompiled instead of compute on every event
type requestShaper struct {
	prefix string
	pr     *urlshaper.Parser
}

// requestShape expects the field passed in to have the form
// VERB /path/of/request HTTP/1.x
// If it does, it will break it apart into components, normalize the URL,
// and add a handful of additional fields based on what it finds.
func (r *requestShaper) requestShape(field string, ev *event.Event,
	options GlobalOptions) {
	if val, ok := ev.Data[field]; ok {
		// start by splitting out method, uri, and version
		parts := strings.Split(val.(string), " ")
		var path string
		if len(parts) == 3 {
			// treat it as METHOD /path HTTP/1.X
			ev.Data[r.prefix+field+"_method"] = parts[0]
			ev.Data[r.prefix+field+"_protocol_version"] = parts[2]
			path = parts[1]
		} else {
			// treat it as just the /path
			path = parts[0]
		}
		// next up, get all the goodies out of the path
		res, err := r.pr.Parse(path)
		if err != nil {
			// couldn't parse it, just pass along the event
			return
		}
		ev.Data[r.prefix+field+"_uri"] = res.URI
		ev.Data[r.prefix+field+"_path"] = res.Path
		if res.Query != "" {
			ev.Data[r.prefix+field+"_query"] = res.Query
		}
		for k, v := range res.QueryFields {
			// only include the keys we want
			if options.RequestParseQuery == "all" ||
				whitelistKey(options.RequestQueryKeys, k) {
				if len(v) > 1 {
					sort.Strings(v)
				}
				ev.Data[r.prefix+field+"_query_"+k] = strings.Join(v, ", ")
			}
		}
		for k, v := range res.PathFields {
			ev.Data[r.prefix+field+"_path_"+k] = v[0]
		}
		ev.Data[r.prefix+field+"_shape"] = res.Shape
		ev.Data[r.prefix+field+"_pathshape"] = res.PathShape
		if res.QueryShape != "" {
			ev.Data[r.prefix+field+"_queryshape"] = res.QueryShape
		}
	}
}

// return true if the key is in the whitelist
func whitelistKey(whiteKeys []string, key string) bool {
	for _, whiteKey := range whiteKeys {
		if key == whiteKey {
			return true
		}
	}
	return false
}

// sendToLibhoney reads from the toBeSent channel and shoves the events into
// libtb events, sending them on their way.
func sendToLibhoney(ctx context.Context, toBeSent chan event.Event, toBeResent chan event.Event,
	delaySending chan int, doneSending chan bool) {
	for {
		// check and see if we need to back off the API because of rate limiting
		select {
		case delay := <-delaySending:
			time.Sleep(time.Duration(delay) * time.Millisecond)
		default:
		}
		// if we have events to retransmit, send those first
		select {
		case ev := <-toBeResent:
			// retransmitted events have already been sampled; always use
			// SendPresampled() for these
			sendEvent(ev)
			continue
		default:
		}
		// otherwise pick something up off the regular queue and send it
		select {
		case ev, ok := <-toBeSent:
			if !ok {
				// channel is closed
				// NOTE: any unrtransmitted retransmittable events will be dropped
				doneSending <- true
				return
			}
			sendEvent(ev)
			continue
		default:
		}
		// no events at all? chill for a sec until we get the next one
		time.Sleep(100 * time.Millisecond)
	}
}

// sendEvent does the actual handoff to libtb
func sendEvent(ev event.Event) {
	if ev.SampleRate == -1 {
		// drop the event!
		logrus.WithFields(logrus.Fields{
			"event": ev,
		}).Debug("droppped event due to sampling")
		return
	}
	libhEv := libtb.NewEvent()
	libhEv.Metadata = ev
	libhEv.Timestamp = ev.Timestamp
	libhEv.SampleRate = uint(ev.SampleRate)
	if err := libhEv.Add(ev.Data); err != nil {
		logrus.WithFields(logrus.Fields{
			"event": ev,
			"error": err,
		}).Error("Unexpected error adding data to libtb event")
	}
	if err := libhEv.SendPresampled(); err != nil {
		logrus.WithFields(logrus.Fields{
			"event": ev,
			"error": err,
		}).Error("Unexpected error event to libtb send")
	}
}

// handleResponses reads from the response queue, logging a summary and debug
// re-enqueues any events that failed to send in a retryable way
func handleResponses(responses chan libtb.Response, stats *responseStats,
	toBeResent chan event.Event, delaySending chan int,
	options GlobalOptions) {
	go logStats(stats, options.StatusInterval)

	for rsp := range responses {
		stats.update(rsp)
		logfields := logrus.Fields{
			"status_code": rsp.StatusCode,
			"body":        strings.TrimSpace(string(rsp.Body)),
			"duration":    rsp.Duration,
			"error":       rsp.Err,
			"timestamp":   rsp.Metadata.(event.Event).Timestamp,
		}
		// if this is an error we should retry sending, re-enqueue the event
		if options.BackOff && (rsp.StatusCode == 429 || rsp.StatusCode == 500) {
			logfields["retry_send"] = true
			delaySending <- 1000 / int(options.NumSenders) // back off for a little bit
			toBeResent <- rsp.Metadata.(event.Event)       // then retry sending the event
		} else {
			logfields["retry_send"] = false
		}
		logrus.WithFields(logfields).Debug("event send record received")
	}
}

// logStats dumps and resets the stats once every minute
func logStats(stats *responseStats, interval uint) {
	logrus.Debugf("Initializing stats reporting. Will print stats once/%d seconds", interval)
	if interval == 0 {
		// interval of 0 means don't print summary status
		return
	}
	ticker := time.NewTicker(time.Second * time.Duration(interval))
	for range ticker.C {
		stats.logAndReset()
	}
}
